src/workerd/io/io-gate.c++

```diff
@@ -62,6 +62,24 @@ kj::Promise<void> InputGate::onBroken() {
   }
 }
 
+InputGate::Lock::Lock(InputGate& gate)
+    : gate(&gate),
+      cs(gate.isCriticalSection
+          ? kj::Maybe(kj::addRef(static_cast<CriticalSection&>(gate)))
+          : nullptr) {
+  InputGate* gateToLock = &gate;
+
+  KJ_IF_MAYBE(c, cs) {
+    if (c->get()->state == CriticalSection::REPARENTED) {
```

Comment:
To summarize these ifs to my understanding:
- If our gate is a critical section AND it has been re-parented, get the effective parent.
- If our gate is a critical section AND it has NOT been re-parented, use the gate.
- If our gate is not a critical section, use the gate.

If you're interested, I could suggest some ways to make this code easier to skim. Otherwise, maybe a comment/empty else branch calling out that `cs` is `gate`?

Reply:
I agree this code is pretty hard to follow. It took me a while to figure it out myself. (And I think I wrote it... though there have been some changes.)

Right now I'm drowning in code reviews so I prefer to merge this as-is and move on.